### ISSUE TITLE
Error listing AppService certificates using fluent api

### ISSUE THREAD

**Issue by hovsepm** [Azure/azure-libraries-for-net#51]
_From @NicolaiPetri on July 11, 2017 7:46_

When fetching the list of certificates for specific subscription I'll get the following exception:
"Unable to deserialize the response."
 Microsoft.Rest.SerializationException

The code that fails is 
`var certificates = azure.AppServices.AppServiceCertificates.List();`

The full exception is here :

```
Exception: Microsoft.Rest.SerializationException: Unable to deserialize the response. ---> Newtonsoft.Json.JsonSerializationException: Cannot deserialize the current JSON array (e.g. [1,2,3]) into type 'Microsoft.Azure.Management.AppService.Fluent.Models.Page'1[Microsoft.Azure.Management.AppService.Fluent.Models.CertificateInner]' because the type requires a JSON object (e.g. {"name":"value"}) to deserialize correctly.
 To fix this error either change the JSON to a JSON object (e.g. {"name":"value"}) or change the deserialized type to an array or a type that implements a collection interface (e.g. ICollection, IList) like List<T> that can be deserialized from a JSON array. JsonArrayAttribute can also be added to the type to force it to deserialize from a JSON array.
 Path '', line 1, position 1.
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EnsureArrayContract(JsonReader reader, Type objectType, JsonContract contract)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateList(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, Object existingValue, String id)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
    at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
    at Microsoft.Rest.Serialization.SafeJsonConvert.DeserializeObject[T](String json, JsonSerializerSettings settings)
    at Microsoft.Azure.Management.AppService.Fluent.CertificatesOperations.<ListWithHttpMessagesAsync>d__5.MoveNext()
    --- End of inner exception stack trace ---
    at Microsoft.Azure.Management.AppService.Fluent.CertificatesOperations.<ListWithHttpMessagesAsync>d__5.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at Microsoft.Azure.Management.AppService.Fluent.CertificatesOperationsExtensions.<ListAsync>d__1.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at Microsoft.Azure.Management.AppService.Fluent.AppServiceCertificatesImpl.<ListInnerAsync>d__4.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at Microsoft.Azure.Management.ResourceManager.Fluent.Core.TopLevelModifiableResources`5.List()
    at Microsoft.Azure.Management.AppService.Fluent.AppServiceCertificatesImpl.Microsoft.Azure.Management.ResourceManager.Fluent.Core.CollectionActions.ISupportsListing<Microsoft.Azure.Management.AppService.Fluent.IAppServiceCertificate>.List()
    at AzureExporter2.ResourceIdCachedService.GetResourceId(IAzure azure, String resourceId, String resourceType, String resourceName) in C:\Projects\Dashboard\azure_exporter\AzureExporter2\ResourceIdCachedService.cs:line 47
    at AzureExporter2.Program.<>c__DisplayClass1_0.<Main>b__1(IAsyncResult ar) in C:\Projects\Dashboard\azure_exporter\AzureExporter2\Program.cs:line 88
```

I can on request provide the failing and valid json response in a private message.

_Copied from original issue: Azure/azure-sdk-for-net#3474_

**Comment by hovsepm** [comment]
@NicolaiPetri Which version of the nuget package are you using?

**Comment by hovsepm** [comment]
_From @NicolaiPetri on July 12, 2017 11:7_

Hi, 
I tested this with both version 1.0.0 and latest 1.1.3 and it's the same problem.
I did a workaround in my code where I loop thru all resource groups and list certificates per group. This works without problems - so it's not a problem with a specific certificate in the subscription. 
I have a theory that this happens when the rest api can't return all certificates in a single response and then a continuation link is given.

**Comment by NicolaiPetri** [comment]
Hi,

As written earlier I tested this with both version 1.0.0 and (at that time)
latest 1.1.3 and it was the same problem

---
Nicolai

**Comment by jeffputz** [comment]
Wow, so first thing I've ever decided to use the SDK for and this is it. How lucky am I? :) The first problem is that the response coming back from Azure isn't the same as the doc:

https://docs.microsoft.com/en-us/rest/api/appservice/certificates/list

It's just an array and not contained against a "value." I was going to do a PR, but if you still want it enclosed in a Page, I'm not sure how you do that because the Items collection is private and only populated via deserialization.

**Comment by jeffputz** [comment]
Yikes, so this is still broken, yes? Is the issue that no one can decide to fix the management API or the SDK?

**Comment by hovsepm** [comment]
@jeffputz the issue comes from wrong swagger spec and actual payload. IT should be fixed on backend side.

**Comment by jeffputz** [comment]
So who owns the problem then?

**Comment by shilsmarty** [comment]
Sorry , for intruding the discussion.

Am getting same error from AppService fluent when calling Get by ResourceGroup 

**Example Code - AppServices.FunctionApps.GetByResourceGroupAsync(ResourceGroupName, resourceIdentifier);**

This is also happening for List by ResourceGroup as well.

Well there is content in the exception but don't want to deserialize it.

Please suggest.

**Comment by eydelrivero** [comment]
Hi. Any update on this?

**Comment by Suchiman** [comment]
Almost two years later and still no solution in sight? This is unacceptable.

**Comment by Suchiman** [comment]
Here's a workaround: 
```csharp
class FixCertificatesHandler : DelegatingHandler
{
    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        HttpResponseMessage resp = await base.SendAsync(request, cancellationToken);
        if (resp.StatusCode == System.Net.HttpStatusCode.OK && request.RequestUri.LocalPath.EndsWith("/providers/Microsoft.Web/certificates"))
        {
            var str = await resp.Content.ReadAsStringAsync();
            str = "{ \"value\": " + str + " }";
            resp.Content.Dispose();
            resp.Content = new StringContent(str);
        }
        return resp;
    }
}
```
and then use `Azure.Configure().WithDelegatingHandler(new FixCertificatesHandler())`

**Comment by Stephanvs** [comment]
> Here's a workaround:
> 
> ```cs
> class FixCertificatesHandler : DelegatingHandler
> {
>     protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
>     {
>         HttpResponseMessage resp = await base.SendAsync(request, cancellationToken);
>         if (resp.StatusCode == System.Net.HttpStatusCode.OK && request.RequestUri.LocalPath.EndsWith("/providers/Microsoft.Web/certificates"))
>         {
>             var str = await resp.Content.ReadAsStringAsync();
>             str = "{ \"value\": " + str + " }";
>             resp.Content.Dispose();
>             resp.Content = new StringContent(str);
>         }
>         return resp;
>     }
> }
> ```
> 
> and then use `Azure.Configure().WithDelegatingHandler(new FixCertificatesHandler())`

I actually ran into the exact same issue as the OP. Interesting though that this issue is already from july 2017! The workaround posted does seem to work though! Thanks @Suchiman

**Comment by Suchiman** [comment]
Another workaround is to use `ListByResourceGroupAsync` instead of just `ListAsync` which is probably more often what you actually want.

**Comment by Suchiman** [comment]
I have updated my workaround due to it breaking the `ListByResourceGroupAsync` method (the url pattern was matching too wildly). @Stephanvs please make sure to update the Workaround or switch to using `ListByResourceGroupAsync` ,which works out of the box, to avoid pulling out your hair 😉

**Comment by weidongxu-microsoft** [comment]
Service API 2018-11-01 had fixed it. I still have question with service on API 2019-08-01 version. When it is solved and we could proceed with https://github.com/Azure/azure-libraries-for-net/issues/857, and that would fix this.

**Comment by weidongxu-microsoft** [comment]
Fixed in 1.30, to be released about 9th Jan.

@Suchiman @Stephanvs 

The workaround would be broken, if updated to fixed version (1.30 or above).

One could additionally match "api-version=2018-02-01" in URL Query, which was the service version having this bug.

1.30 is upgraded to "api-version=2019-08-01".
